File: system-design/oop-patterns/05-abstract-factory-pattern/full/java/Main.java
import theme.UIFactory;
import theme.LightThemeFactory;
import theme.DarkThemeFactory;

public class Main {
    static void renderScreen(UIFactory factory) {
        System.out.println(factory.createButton().render());
        System.out.println(factory.createLabel().render());
    }

    public static void main(String[] args) {
        System.out.println("Light theme:");
        renderScreen(new LightThemeFactory());

        System.out.println("Dark theme:");
        renderScreen(new DarkThemeFactory());
    }
}

File: system-design/oop-patterns/05-abstract-factory-pattern/full/ts/main.ts
import type { UIFactory } from './theme/UIFactory';
import { LightThemeFactory } from './theme/LightThemeFactory';
import { DarkThemeFactory } from './theme/DarkThemeFactory';

function renderScreen(factory: UIFactory): void {
  console.log(factory.createButton().render());
  console.log(factory.createLabel().render());
}

console.log('Light theme:');
renderScreen(new LightThemeFactory());

console.log('Dark theme:');
renderScreen(new DarkThemeFactory());
